docs/source/expr-system-internals.rst

```diff
@@ -0,0 +1,206 @@
+
+To address this, each expression computes and caches its name and token upon construction. These values are stored and serialized to ensure pickle roundtrip stability. The token is accessible via ``_determ_token`` or ``deterministic_token``.
+
+Caching and Singletons
+^^^^^^^^^^^^^^^^^^^^^^
+
+Despite efforts to keep expressions stateless, in practice many attributes are computed on demand and cached via ``functools.cached_property``. This defers computation but complicates reasoning about when and how state is evaluated.
+
+Cached properties are typically serialized with the expression, unless ``_pickle_functools_cache`` is set to False.
+
+To preserve cache values during repeated optimization (which recreates expressions), most classes subclass :class:`^dask.dataframe._expr.SingletonExpr`. This ensures that instances with the same name return a previously created, cached version. This makes expressions effectively immutable singletons — and must not be mutated in-place.
+
+Optimization Procedure
+----------------------
+
+Expressions form a directed graph structure: when one expression is passed as an operand to another, it becomes a dependency. While this starts as a tree, deduplication by name quickly transforms it into a directed acyclic graph (DAG) — a key property for optimization.
+
+The optimizer currently performs the following five steps:
+
+* Simplify
+
+* Rewrite / Tune
+
+* Lower
+
+* Simplify (again)
+
+* Fuse
+
+Simplify
+^^^^^^^^
+
+Simplification rewrites expressions into more optimal but semantically equivalent forms. A common example is pushing projections or filters down the graph to reduce computation earlier.
+
+Constraints for simplification (not enforced at runtime):
+
+* The number of partitions (npartitions) must not increase.
+
+* No computations with side effects (e.g., computing divisions) may occur.
+
+Rewrite / Tune
+^^^^^^^^^^^^^^
+
+This step implements performance tuning based on heuristics. Typically this targets a more efficient intermediate partitioning.
+
+Two examples:
+
+* Fuse I/O operations based on column projection (e.g., ``FusedIO``)
+
+* Choose an appropriate ``split_out`` to balance partitioning
+
+This step does not alter the logical meaning of expressions but adjusts execution-related parameters.
+
+Lowering
+^^^^^^^^
+
+At this stage, abstract operations are transformed into concrete execution strategies.
+
+For example:
+
+A logical ``Merge`` might become a ``BlockwiseMerge`` if the input DataFrames are already partitioned appropriately.
+
+In less favorable cases, a general ``HashJoinP2P`` might be selected instead.
+
+This marks the transition from logical to physical plan, akin to traditional query planners.
+
+After lowering, a second simplification step is applied to the resulting expressions.
+
+Fuse
+^^^^
+
+Linear chains of blockwise tasks are combined into a single task, minimizing scheduler overhead.
```

Comment:
I think that this could use a brief example.  It's not immediately clear to me. 

Also, I thought that there was a global pass around deduplication.  Is that no longer happening?

Reply:
I don't understand what you are referring to. Nothing is explicitly deduplicated since everything is implicitly deduplicated by the tokens.

Reply:
Yeah, not actual deduplication at the node level, but things like reading the same parquet file twice, after we push down columns onto a read_parquet expression.  My understanding is that there was a non-traversal optimization pass.  Something that looked at things globally (my knowledge is out of date though).

Reply:
there is no global optimization step. this is all down with simplify_up